internal/controller/notificationservice_controller.go

```diff
@@ -78,20 +80,18 @@ func (r *NotificationServiceReconciler) Reconcile(ctx context.Context, req ctrl.
 		results, err := GetResultsFromPipelineRun(pipelineRun)
 		if err != nil {
 			logger.Error(err, "Failed to get results for pipelineRun ", pipelineRun.Name)
-		} else {
-			fmt.Printf("Results for pipelinerun %s are: %s\n", pipelineRun.Name, results)
-			err = AddAnnotationToPipelineRun(ctx, pipelineRun, r, NotificationPipelineRunAnnotation, NotificationPipelineRunAnnotationValue)
-			if err != nil {
-				logger.Error(err, "Failed to add annotation")
-			}
+			return ctrl.Result{}, err
+		}
+		fmt.Printf("Results for pipelinerun %s are: %s\n", pipelineRun.Name, results)
+		err = AddAnnotationToPipelineRun(ctx, pipelineRun, r, NotificationPipelineRunAnnotation, NotificationPipelineRunAnnotationValue)
+		if err != nil {
+			logger.Error(err, "Failed to add annotation to pipelinerun ", pipelineRun.Name)
+			return ctrl.Result{}, err
 		}
-	}
-
-	if IsPipelineRunEndedSuccessfully(pipelineRun) &&
-		IsAnnotationExistInPipelineRun(pipelineRun, NotificationPipelineRunAnnotation, NotificationPipelineRunAnnotationValue) {
 		err = RemoveFinalizerFromPipelineRun(ctx, pipelineRun, r, NotificationPipelineRunFinalizer)
```

Comment:
I wonder if we can have a case where we will set the annotation but then fail to remove the finalizer. If it will happen the pipelinerun would never get deleted. I think that we should have another predicate that will invoke reconcile for completed pipelines that has the finalizer on them.

Reply:
we should also add a test case for such scenario. One of the worst things we can do is to block the deletion of pipelineruns. It can cause an outage for useres.